core/executor/src/wasm_executor.rs

```diff
@@ -214,44 +241,44 @@ impl_function_executor!(this: FunctionExecutor<'e, E>,
 		Ok(())
 	},
 	ext_exists_storage(key_data: *const u8, key_len: u32) -> u32 => {
-		let key = this.memory.get(key_data, key_len as usize).map_err(|_| UserError("Invalid attempt to determine key in ext_exists_storage"))?;
+		let key = this.memory.get(key_data, key_len as usize)
+			.map_err(|_| UserError("Invalid attempt to determine key in ext_exists_storage"))?;
 		Ok(if this.ext.exists_storage(&key) { 1 } else { 0 })
 	},
-	ext_exists_child_storage(storage_key_data: *const u8, storage_key_len: u32, key_data: *const u8, key_len: u32) -> u32 => {
-		let storage_key = this.memory.get(
-			storage_key_data,
-			storage_key_len as usize
-		).map_err(|_| UserError("Invalid attempt to determine storage_key in ext_exists_child_storage"))?;
-		let key = this.memory.get(key_data, key_len as usize).map_err(|_| UserError("Invalid attempt to determine key in ext_exists_child_storage"))?;
-		let storage_key = ChildStorageKey::from_vec(storage_key)
-			.ok_or_else(||
-				UserError("ext_exists_child_storage: child storage key is not valid")
-			)?;
-		Ok(if this.ext.exists_child_storage(storage_key, &key) { 1 } else { 0 })
+	ext_exists_child_storage(
+		keyspace_data: *const u8,
+		keyspace_len: u32,
+		root_data: *const u8,
+		root_len: u32,
+		key_data: *const u8,
+		key_len: u32) -> u32 => {
+		let keyspace = this.memory.get(keyspace_data, keyspace_len as usize)
+			.map_err(|_| UserError("Invalid attempt to determine storage_key in ext_set_child_storage"))?;
+		let root = this.memory.get(root_data, root_len as usize)
+			.map_err(|_| UserError("Invalid attempt to determine storage_key in ext_set_child_storage"))?;
+		let key = this.memory.get(key_data, key_len as usize)
+			.map_err(|_| UserError("Invalid attempt to determine key in ext_exists_child_storage"))?;
+		let root = if root.len() > 0 { Some(&root[..]) } else { None };
```

Comment:
isn't the this.memory.get should be performed inside the yes branch ? Because root_data shouldn't be read in case len is 0 no ?

Reply:
definitely